temporalio/ext/src/client.rs

```diff
@@ -0,0 +1,314 @@
+use std::{collections::HashMap, future::Future, time::Duration};
+
+use temporal_client::{
+    ClientInitError, ClientKeepAliveConfig, ClientOptionsBuilder, ClientTlsConfig,
+    ConfiguredClient, HttpConnectProxyOptions, RetryClient, RetryConfig,
+    TemporalServiceClientWithMetrics, TlsConfig, WorkflowService,
+};
+
+use magnus::{
+    block::Proc, class, function, method, prelude::*, scan_args, value::Opaque, DataTypeFunctions,
+    Error, RString, Ruby, TypedData, Value,
+};
+use tonic::{metadata::MetadataKey, Status};
+use url::Url;
+
+use super::{error, new_error, ROOT_MOD};
+use crate::{
+    runtime::{Runtime, RuntimeHandle},
+    util::Struct,
+};
+use std::str::FromStr;
+
+const SERVICE_WORKFLOW: u8 = 1;
+const SERVICE_OPERATOR: u8 = 2;
+const SERVICE_CLOUD: u8 = 3;
+const SERVICE_TEST: u8 = 4;
+const SERVICE_HEALTH: u8 = 5;
+
+pub fn init(ruby: &Ruby) -> Result<(), Error> {
+    let root_mod = ruby.get_inner(&ROOT_MOD);
+
+    let class = root_mod.define_class("Client", class::object())?;
+    class.const_set("SERVICE_WORKFLOW", SERVICE_WORKFLOW)?;
+    class.const_set("SERVICE_OPERATOR", SERVICE_OPERATOR)?;
+    class.const_set("SERVICE_CLOUD", SERVICE_CLOUD)?;
+    class.const_set("SERVICE_TEST", SERVICE_TEST)?;
+    class.const_set("SERVICE_HEALTH", SERVICE_HEALTH)?;
+    class.define_singleton_method("async_new", function!(Client::async_new, 2))?;
+    class.define_method("async_invoke_rpc", method!(Client::async_invoke_rpc, -1))?;
+
+    let inner_class = class.define_class("RpcFailure", class::object())?;
+    inner_class.define_method("code", method!(RpcFailure::code, 0))?;
+    inner_class.define_method("message", method!(RpcFailure::message, 0))?;
+    inner_class.define_method("details", method!(RpcFailure::details, 0))?;
+    Ok(())
+}
+
+type CoreClient = RetryClient<ConfiguredClient<TemporalServiceClientWithMetrics>>;
+
+#[derive(DataTypeFunctions, TypedData)]
+#[magnus(class = "Temporalio::Internal::Bridge::Client", free_immediately)]
+pub struct Client {
+    pub(crate) core: CoreClient,
+    runtime_handle: RuntimeHandle,
+}
+
+macro_rules! rpc_call {
+    ($client:ident, $block:ident, $call:ident, $call_name:ident) => {{
+        if $call.retry {
+            let mut core_client = $client.core.clone();
+            let req = $call.into_request()?;
+            rpc_resp(
+                $client,
+                $block,
+                async move { core_client.$call_name(req).await },
+            )
+        } else {
+            let mut core_client = $client.core.clone().into_inner();
+            let req = $call.into_request()?;
+            rpc_resp(
+                $client,
+                $block,
+                async move { core_client.$call_name(req).await },
+            )
+        }
+    }};
+}
+
+impl Client {
+    pub fn async_new(ruby: &Ruby, runtime: &Runtime, options: Struct) -> Result<(), Error> {
```

Comment:
So notice how we have done passing values from Ruby to Rust. I originally used a massive kwargs list via https://docs.rs/magnus/latest/magnus/scan_args/fn.get_kwargs.html, but not only is that limited, using keyword-based struct arguments helps us eagerly catch if we're using the wrong key. For performance I will probably change to use symbols instead of Rust strings for accessing struct fields (which become Ruby strings on each ask).

Reply:
Changed all member accesses on struct to be lazy-static symbols